packages/react/src/components/Copy/Copy.js

```diff
@@ -6,91 +6,101 @@
  */
 
 import PropTypes from 'prop-types';
-import React, { Component } from 'react';
+import React, { useState, useEffect, useCallback } from 'react';
+import debounce from 'lodash/debounce';
 import classnames from 'classnames';
 import { settings } from 'carbon-components';
 
 const { prefix } = settings;
 
-export default class Copy extends Component {
-  static propTypes = {
-    /**
-     * Pass in content to be rendred in the underlying <button>
-     */
-    children: PropTypes.node,
-
-    /**
-     * Specify an optional className to be applied to the underlying <button>
-     */
-    className: PropTypes.string,
-
-    /**
-     * Specify the string that is displayed when the button is clicked and the
-     * content is copied
-     */
-    feedback: PropTypes.string,
-
-    /**
-     * Specify the time it takes for the feedback message to timeout
-     */
-    feedbackTimeout: PropTypes.number,
-
-    /**
-     * Specify an optional `onClick` handler that is called when the underlying
-     * <button> is clicked
-     */
-    onClick: PropTypes.func,
-  };
-
-  static defaultProps = {
-    feedback: 'Copied!',
-    feedbackTimeout: 2000,
-    onClick: () => {},
+export default function Copy({
+  children,
+  className,
+  feedback,
+  feedbackTimeout,
+  onClick,
+  ...other
+}) {
+  const [animation, setAnimation] = useState('');
+  const classNames = classnames(className, {
+    [`${prefix}--copy-btn--animating`]: animation,
+    [`${prefix}--copy-btn--${animation}`]: animation,
+  });
+  const handleFadeOut = useCallback(
```

Comment:
Would it make sense for this work to be in an effect parameterized by `animation` changing? Something like:

```jsx
useEffect(() => {
  if (animation !== 'fade-in') {
    return;
  }

  const timeoutId = window.setTimeout(() => {
    setAnimation('fade-out');
  }, [feedbackTimeout]);

  return () => {
    window.clearTimeout(timeoutId);
  };
}, [animation, feedbackTimeout]);
```

Reply:
does this eliminate our debouncing then? I moved towards using debounce to eliminate the need for managing timers and refs

Reply:
Got it, is `handleFadeOut` called multiple time which is why we need debounce here?

Reply:
yeah for both the click handler and fadeout handler it's for cleanup purposes

Reply:
yeah the tooltip needs to persist if the button is clicked again and should fade out after the timeout expires for the most recent click of the button. with `useEffect` all clicks following the first click are ignored until the tooltip fades out (based on the timeout from the very first click). so we debounce the click and fadeout handlers because they can fire multiple times

Reply:
Hmm, this seems like a use-case we could achieve with `useEffect`, right? What would make the clicks become ignored? It seemed like on click we want to update the animation state and then enqueue a timeout, cancelling it if the button is clicked and creating a new one. Is there something else that I'm missing here? 👀

Reply:
if the timeout management occurs in `useEffect` doesn't that require another render? patching with `useEffect` results in `clearTimeout` firing after the timeout has already completed

![gif](https://user-images.githubusercontent.com/8265238/72167773-6a9a6800-3391-11ea-9ebf-bab39a0defe1.gif)